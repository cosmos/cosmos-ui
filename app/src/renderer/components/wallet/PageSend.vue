<template lang='pug'>
tm-page(title='Send')
  div(slot="menu"): tool-bar
  tm-form-struct(:submit="onSubmit")
    tm-part(title='Denomination Options')
      tm-form-group(:error='$v.fields.denom.$error'
        field-id='send-denomination' field-label='Denomination')
        tm-field#send-denomination(
          type="select"
          v-model="fields.denom"
          :options="denominations"
          placeholder="Select token...")
        tm-form-msg(name='Denomination' type='required' v-if='!$v.fields.denom.required')

    tm-part(title='Transaction Details')
      tm-form-group(:error='$v.fields.zoneId.$error'
        v-if="config.devMode"
        field-id='send-zone-id' field-label='Zone ID')
        tm-field#send-zone-id(
          type="select"
          v-model="fields.zoneId"
          :options="zoneIds"
          placeholder="Select zone...")
        tm-form-msg(name='Zone' type='required' v-if='!$v.fields.zoneId.required')
      tm-form-group(:error='$v.fields.address.$error'
        field-id='send-address' field-label='Send To')
        tm-field-group
          tm-field#send-address(
            type='text'
            v-model='fields.address'
            placeholder='Address')
        tm-form-msg(name='Address' type='required' v-if='!$v.fields.address.required')
        tm-form-msg(name='Address' type='bech32' :body="bech32error" v-else-if='!$v.fields.address.bech32Validate')

      tm-form-group(:error='$v.fields.amount.$error'
        field-id='send-amount' field-label='Amount')
        tm-field-group
          tm-field#send-amount(
            type='number'
            v-model='fields.amount'
            placeholder='Amount')
        tm-form-msg(name='Amount' type='required' v-if='!$v.fields.amount.required')
        tm-form-msg(name='Amount' type='between' :min='max ? 1 : 0' :max='max'
          v-if='!$v.fields.amount.between')

      p(v-if='mockedConnector')
        span Try sending to the address "
        strong(style="font-weight: bold") cosmosaccaddr1p6zajjw6xged056andyhn62lm7axwzyspkzjq0
        span ", it's a friendly bot which will send the money back to you!
      br(v-if='mockedConnector')

    div(slot='footer')
      div
      tm-btn(v-if='sending' value='Sending...' disabled color="primary")
      tm-btn(v-else @click='onSubmit' value="Send Tokens" color="primary")

  tm-modal-send-confirmation(v-if="confirmationPending" @approved="onApproved" @canceled="onCancel" :amount="fields.amount" :recipient="fields.address" :denom="fields.denom")
</template>

<script>
import b32 from "scripts/b32"
import { required, between } from "vuelidate/lib/validators"
import { mapActions, mapGetters } from "vuex"
import {
  TmBtn,
  TmFieldGroup,
  TmFormGroup,
  TmFormStruct,
  TmPage,
  TmPart,
  TmField,
  TmFormMsg
} from "@tendermint/ui"

import FieldAddon from "common/TmFieldAddon"
import ToolBar from "common/TmToolBar"
import TmModalSendConfirmation from "wallet/TmModalSendConfirmation"
export default {
  components: {
    TmBtn,
    TmField,
    FieldAddon,
    TmFieldGroup,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    TmPage,
    TmPart,
    ToolBar,
    TmModalSendConfirmation
  },
  computed: {
    ...mapGetters(["wallet", "lastHeader", "config", "mockedConnector"]),
    max() {
      let denom = this.wallet.balances.find(b => b.denom === this.denom)
      return (denom && denom.amount) || 0
    },
    denominations() {
      return this.wallet.balances.map(i => ({
        key: i.denom.toUpperCase(),
        value: i.denom
      }))
    },
    zoneIds() {
      return this.wallet.zoneIds.map(z => ({ key: z, value: z }))
    }
  },
  data: () => ({
    bech32error: null,
    fields: {
      address: "",
      amount: null,
      denom: "",
      zoneId: "cosmos-hub-1"
    },
    confirmationPending: false,
    sending: false
  }),
  methods: {
    resetForm() {
      this.fields.address = ""
      this.fields.amount = null
      this.sending = false
      this.$v.$reset()
    },
    onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return

      this.confirmationPending = true
    },
    async onApproved() {
      this.confirmationPending = false
      this.sending = true
      let amount = +this.fields.amount
      let address = this.fields.address
      let denom = this.fields.denom
      try {
        // if address has a slash, it is IBC address format
        let type = "send"
        // TODO reenable when we have IBC
        // if (this.lastHeader.chain_id !== zoneId) {
        //   type = "ibcSend"
        //   address = `${zoneId}/${address}`
        // } else {
        //   type = "send"
        // }
        await this.sendTx({
          type,
          to: address,
          amount: [{ denom, amount: amount.toString() }]
        })
        this.sending = false
        this.$store.commit("notify", {
          title: "Successfully Sent",
          body: `Successfully sent ${amount} ${denom.toUpperCase()} to ${address}`
        })
        // resets send transaction form
        this.resetForm()
        // refreshes user transaction history
        this.$store.dispatch("queryWalletHistory")
      } catch (err) {
        this.sending = false
        this.$store.commit("notifyError", {
          title: "Error Sending",
          body: `An error occurred while trying to send: "${err.message}"`
        })
      }
    },
    onCancel() {
      this.confirmationPending = false
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        this.bech32error = null
        return true
      } catch (error) {
        this.bech32error = error.message
        return false
      }
    },
    ...mapActions(["sendTx"])
  },
  props: ["denom"],
  mounted() {
    if (this.denom) {
      this.fields.denom = this.denom
    }
    this.fields.zoneId = this.wallet.zoneIds[0]
  },
  validations() {
    return {
      fields: {
        address: {
          required,
          bech32Validate: this.bech32Validate
        },
        amount: {
          required,
          between: between(1, this.max)
        },
        denom: { required },
        zoneId: { required }
      }
    }
  },
  watch: {
    // TODO ignored while we don't have IBC
    // // if the zoneId gets added at a later time
    // "wallet.zoneIds": () => {
    //   this.fields.zoneId = this.wallet.zoneIds[0]
    // }
  }
}
</script>
