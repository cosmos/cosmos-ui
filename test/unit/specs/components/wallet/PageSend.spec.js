import setup from "../../../helpers/vuex-setup"
import Vuelidate from "vuelidate"
import PageSend from "renderer/components/wallet/PageSend"

describe("PageSend", () => {
  let wrapper, store, node

  let { mount, localVue } = setup()
  localVue.use(Vuelidate)

  beforeEach(async () => {
    let test = mount(PageSend, {
      propsData: {
        denom: "fermion"
      }
    })
    wrapper = test.wrapper
    store = test.store
    node = test.node
    store.commit("setAccounts", [
      {
        address: "DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B",
        name: "default",
        password: "1234567890"
      }
    ])
    await store.dispatch("signIn", {
      account: "default",
      password: "1234567890"
    })
    store.commit("setWalletBalances", [
      {
        denom: "mycoin",
        amount: 1000
      },
      {
        denom: "fermion",
        amount: 2300
      }
    ])
    store.commit("setNonce", 1)
  })

  it("has the expected html structure", () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it("should populate the select options with denoms", () => {
    expect(
      wrapper
        .findAll("option")
        .at(0)
        .text()
    ).toBe("Select token...")
    expect(
      wrapper
        .findAll("option")
        .at(1)
        .text()
    ).toBe("MYCOIN")
    expect(
      wrapper
        .findAll("option")
        .at(2)
        .text()
    ).toBe("FERMION")
  })

  it("should populate the select options with networks", () => {
    expect(
      wrapper
        .findAll("option")
        .at(3)
        .text()
    ).toBe("Select zone...")
    expect(
      wrapper
        .findAll("option")
        .at(4)
        .text()
    ).toBe("basecoind-demo1")
  })

  it("should work without providing a default denom", () => {
    let { wrapper } = mount(PageSend)
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it("should show notification for successful send", async () => {
    wrapper.setData({
      fields: {
        denom: "mycoin",
        address: "CE456B8BA9AFD1CBDF4ED14558E8C30691E549EA",
        amount: 2,
        zoneId: "cosmos-hub-1"
      }
    })
    await wrapper.vm.onSubmit()
    // walletSend is async so we need to wait until it is resolved
    expect(store.commit).toHaveBeenCalledWith("notify", expect.any(Object))
  })

  it("should show notification for unsuccessful send", async () => {
    wrapper.setData({
      fields: {
        denom: "notmycoin",
        address: "CE456B8BA9AFD1CBDF4ED14558E8C30691E549EA",
        amount: 2,
        zoneId: "cosmos-hub-1"
      }
    })
    node.sign = () => Promise.reject()
    await wrapper.vm.onSubmit()
    expect(store.state.notifications.length).toBe(1)
    expect(store.state.notifications[0].title).toBe("Error Sending")
    expect(store.state.notifications[0]).toMatchSnapshot()
  })
})
