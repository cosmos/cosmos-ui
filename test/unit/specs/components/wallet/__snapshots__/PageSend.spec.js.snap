// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSend has the expected html structure 1`] = `
<div
  class="ni-page"
>
  <header
    class="ni-page-header"
  >
    <div
      class="ni-page-header-container"
    >
      <div
        class="ni-page-header-text"
      >
        <div
          class="ni-page-header-title"
        >
          <h2>
            Send
          </h2>
        </div>
        <div
          class="ni-page-header-subtitle"
        >
          <h3>
            
          </h3>
        </div>
      </div>
      <menu
        class="ni-page-header-menu"
      >
        <div>
          <div>
            <div
              class="ni-tool-bar"
            >
              <div
                class="ni-tool-bar-container"
              >
                <div
                  class="main"
                >
                  <a
                    class="router-link-exact-active router-link-active"
                    href="#/"
                  >
                    <i
                      class="material-icons"
                    >
                      arrow_back
                    </i>
                    <div
                      class="label"
                    >
                      Back
                    </div>
                  </a>
                </div>
                <a
                  class="help"
                >
                  <i
                    class="material-icons"
                  >
                    help_outline
                  </i>
                </a>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
  </header>
  <main
    class="ni-page-main"
  >
    <form
      class="ni-form"
    >
      <div
        class="ni-form-container"
      >
        <!---->
        <main
          class="ni-form-main"
        >
          <section
            class="ni-part"
          >
            <div
              class="ni-part-container"
            >
              <header
                class="ni-part-header"
              >
                <div
                  class="ni-part-header-container"
                >
                  <div
                    class="ni-part-title h5"
                  >
                    Denomination Options
                  </div>
                  <menu
                    class="ni-part-menu"
                  />
                </div>
              </header>
              <main
                class="ni-part-main"
              >
                <div
                  class="ni-form-group"
                >
                  <!---->
                  <label
                    class="ni-form-group__label"
                    for="send-denomination"
                  >
                    Denomination
                  </label>
                  <div
                    class="ni-form-group__field"
                  >
                    <div
                      class="ni-select"
                      id="send-denomination"
                    >
                      <select
                        class="ni-field ni-field-select"
                      >
                        <option
                          disabled="disabled"
                          hidden="hidden"
                          selected="selected"
                          value=""
                        >
                          Select token...
                        </option>
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <!---->
                        <option
                          value="ATOM"
                        >
                          ATOM
                        </option>
                        <option
                          value="FERMION"
                        >
                          FERMION
                        </option>
                      </select>
                      <div
                        class="ni-field-select-addon"
                      >
                        <i
                          class="material-icons"
                        >
                          arrow_drop_down
                        </i>
                      </div>
                    </div>
                    <div
                      class="ni-form-msg sm ni-form-msg--error"
                    >
                      Denomination is required
                    </div>
                  </div>
                </div>
              </main>
            </div>
          </section>
          <section
            class="ni-part"
          >
            <div
              class="ni-part-container"
            >
              <header
                class="ni-part-header"
              >
                <div
                  class="ni-part-header-container"
                >
                  <div
                    class="ni-part-title h5"
                  >
                    Transaction Details
                  </div>
                  <menu
                    class="ni-part-menu"
                  />
                </div>
              </header>
              <main
                class="ni-part-main"
              >
                <div
                  class="ni-form-group"
                >
                  <!---->
                  <label
                    class="ni-form-group__label"
                    for="send-address"
                  >
                    Send To
                  </label>
                  <div
                    class="ni-form-group__field"
                  >
                    <div
                      class="ni-field-group"
                    >
                      <input
                        class="ni-field"
                        id="send-address"
                        placeholder="Address"
                        type="text"
                      />
                    </div>
                    <div
                      class="ni-form-msg sm ni-form-msg--error"
                    >
                      Address is required
                    </div>
                    <!---->
                    <!---->
                  </div>
                </div>
                <div
                  class="ni-form-group"
                >
                  <!---->
                  <label
                    class="ni-form-group__label"
                    for="send-amount"
                  >
                    Amount
                  </label>
                  <div
                    class="ni-form-group__field"
                  >
                    <div
                      class="ni-field-group"
                    >
                      <input
                        class="ni-field"
                        id="send-amount"
                        placeholder="Amount"
                        type="number"
                      />
                    </div>
                    <div
                      class="ni-form-msg sm ni-form-msg--error"
                    >
                      Amount is required
                    </div>
                    <!---->
                  </div>
                </div>
              </main>
            </div>
          </section>
        </main>
        <footer
          class="ni-form-footer"
        >
          <div>
            <div />
            <button
              class="ni-btn"
            >
              <span
                class="ni-btn__container ni-btn--primary"
              >
                <!---->
                <!---->
                <span
                  class="ni-btn__value"
                >
                  Send Tokens
                </span>
              </span>
            </button>
          </div>
        </footer>
      </div>
    </form>
  </main>
</div>
`;

exports[`PageSend should show notification for unsuccessful send 1`] = `
Object {
  "body": "An error occurred while trying to send: \\"Error sending transaction\\"",
  "icon": "error",
  "layout": "alert",
  "time": 1608,
  "title": "Error Sending",
  "type": "error",
}
`;
